src/storage/s3.rs

```diff
@@ -0,0 +1,244 @@
+use futures::Future;
+use postgres::Connection;
+use rusoto_core::region::Region;
+use rusoto_credential::DefaultCredentialsProvider;
+use rusoto_s3::{GetObjectRequest, PutObjectRequest, S3Client, S3};
+use std::convert::TryInto;
+use std::io::Read;
+use time::Timespec;
+
+pub(crate) struct S3Backend<'a> {
+    client: S3Client,
+    bucket: &'a str,
+}
+
+impl<'a> S3Backend<'a> {
+    pub(crate) fn new(client: S3Client, bucket: &'a str) -> Self {
+        Self { client, bucket }
+    }
+
+    pub(super) fn get(&self, path: &str) -> Result<Blob, Error> {
+        let res = self
+            .client
+            .get_object(GetObjectRequest {
+                bucket: self.bucket.to_string(),
+                key: path.into(),
+                ..Default::default()
+            })
+            .sync()?;
+
+        let mut b = res.body.unwrap().into_blocking_read();
+        let mut content = Vec::with_capacity(
+            res.content_length
+                .and_then(|l| l.try_into().ok())
+                .unwrap_or(0),
+        );
+        b.read_to_end(&mut content).unwrap();
+
+        let date_updated = parse_timespec(&res.last_modified.unwrap())?;
+
+        Ok(Blob {
+            path: path.into(),
+            mime: res.content_type.unwrap(),
+            date_updated,
+            content,
+        })
+    }
+
+    pub(super) fn store_batch(&self, batch: &[Blob]) -> Result<(), Error> {
+        let mut rt = tokio::runtime::Runtime::new().unwrap();
+        let mut attempts = 0;
+
+        loop {
+            let mut futures = Vec::new();
+            for blob in batch {
+                futures.push(
+                    self.client
+                        .put_object(PutObjectRequest {
+                            bucket: self.bucket.to_string(),
+                            key: blob.path.clone(),
+                            body: Some(blob.content.clone().into()),
+                            content_type: Some(blob.mime.clone()),
+                            ..Default::default()
+                        })
+                        .inspect(|_| {
+                            crate::web::metrics::UPLOADED_FILES_TOTAL.inc_by(1);
+                        }),
+                );
+            }
+            attempts += 1;
+
+            match rt.block_on(::futures::future::join_all(futures)) {
+                // this batch was successful, start another batch if there are still more files
+                Ok(_) => break,
+                Err(err) => {
+                    error!("failed to upload to s3: {:?}", err);
+                    // if a futures error occurs, retry the batch
+                    if attempts > 2 {
+                        panic!("failed to upload 3 times, exiting");
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(not(test))]
+const TIME_FMT: &str = "%a, %d %b %Y %H:%M:%S %Z";
+#[cfg(test)]
+pub(crate) const TIME_FMT: &str = "%a, %d %b %Y %H:%M:%S %Z";
```

Comment:
Couldn't this just be one thing? The only change between test and non-test is the access level, is that that large of an issue?

Reply:
I guess it's fine to always have it `pub(crate)`, yeah.